added recipe for catmacs

pymaximus opened the pull request:
### Brief summary of what the package does

Provides CAT client for Yaesu FT991A Transceiver

### Direct link to the package repository

https://bitbucket.org/pymaximus/catmacs

### Your association with the package

Author/Developer/Maintainer

### Relevant communications with the upstream package maintainer

**None needed** 

### Checklist

Please confirm with `x`:

- [ x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses). 
- [x ] I've read [CONTRIBUTING.md](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.md)
- [ x] I've used [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x ] My elisp byte-compiles cleanly
- [ x] `M-x checkdoc` is happy with my docstrings
- [ x] I've built and installed the package using the instructions in [CONTRIBUTING.md](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.md)

alphapapa commented:
Hey there,

I can't speak to the package's intended purpose, but it looks neat!  Here's a little feedback:

* Not to discourage you, but you say that it's a prototype, so I'm not sure it belongs on MELPA yet.  Not that there are any hard rules here, though.
* You probably shouldn't bind `C-c m` in the global keymap when the package is loaded.  Better to define a global minor mode that activates and deactivates the keybindings.
* I don't know anything about the CAT protocol, so: is this package specific to that transceiver, or can it work with anything that uses CAT?  If it's specific to that transceiver, maybe it should be renamed appropriately.  `catmacs` brings to mind `cat for Emacs`, which could be feline in nature, or be related to `cat`, maybe something like `eless`.  :)

Thanks.

pymaximus commented:
Hi there,
Thanks for the feedback.

- I no longer consider this a prototype (although like most projects, it will improve over time). Updated the README 
  To reflect this.
- I removed the global key bindings, good point.
- I still like the sound of catmacs, and would be happy to evolve this over time to handle other compatible rigs.

Cheers / Frank


> On 4 Aug 2017, at 7:55 pm, alphapapa <notifications@github.com> wrote:
> 
> Hey there,
> 
> I can't speak to the package's intended purpose, but it looks neat! Here's a little feedback:
> 
> Not to discourage you, but you say that it's a prototype, so I'm not sure it belongs on MELPA yet. Not that there are any hard rules here, though.
> You probably shouldn't bind C-c m in the global keymap when the package is loaded. Better to define a global minor mode that activates and deactivates the keybindings.
> I don't know anything about the CAT protocol, so: is this package specific to that transceiver, or can it work with anything that uses CAT? If it's specific to that transceiver, maybe it should be renamed appropriately. catmacs brings to mind cat for Emacs, which could be feline in nature, or be related to cat, maybe something like eless. :)
> Thanks.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/melpa/melpa/pull/4912#issuecomment-320209075>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AA-yZyOzRgiATvLjz4LKjYUsNugxCKu_ks5sUuqegaJpZM4OsR_T>.
>

purcell commented:
Thanks. Quick feedback:
- The git URL in the recipe contains `pymaximums@`, and it shouldn't. That's not the proper public git URL for your project.
- `:files` spec is unnecessary and should be removed from the recipe.
- `C-c m` is [reserved for end-user customisation](https://www.gnu.org/software/emacs/manual/html_node/elisp/Key-Binding-Conventions.html) and should not be used. Instead, choose an allowed alternative or (better) provide a prefix map, and tell the user how to bind it to their prefix key of choice.

pymaximus commented:
Thanks for the feedback.

Made updates on the recipe. That should be reflected in the pull request now.

I will look at the prefix map idea (have not used that before). 

Cheers / Frank

> On 8 Aug 2017, at 9:56 am, Steve Purcell <notifications@github.com> wrote:
> 
> Thanks. Quick feedback:
> 
> The git URL in the recipe contains pymaximums@, and it shouldn't. That's not the proper public git URL for your project.
> :files spec is unnecessary and should be removed from the recipe.
> C-c m is reserved for end-user customisation <https://www.gnu.org/software/emacs/manual/html_node/elisp/Key-Binding-Conventions.html> and should not be used. Instead, choose an allowed alternative or (better) provide a prefix map, and tell the user how to bind it to their prefix key of choice.
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/melpa/melpa/pull/4912#issuecomment-320811893>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AA-yZ5JsDcUXa_fpnZh08h_BFz7rhN1Sks5sV6QsgaJpZM4OsR_T>.
>

pymaximus commented:
Hi,

Following up on the keymap and prefix idea (which is rather new to me)..

>> 	• C-c m is reserved for end-user customisation and should not be used. Instead, choose an allowed alternative or (better) provide a prefix map, and tell the user how to bind it to their prefix key of choice.

I defined a prefix map as follows ..

(define-minor-mode catmacs-mode
  "CAT client for Yaesu FT991A Transceiver"
  :lighter " catmacs"
  :keymap (let ((catmacs-keymap (make-sparse-keymap)))
            ;; prefix command
            (define-prefix-command 'catmacs-keymap). <======***
            ;; keys
            (define-key catmacs-keymap (kbd "f") 'catmacs-fa-set)
            (define-key catmacs-keymap (kbd "m") 'catmacs-md-set)
              <snip>


I can invoke the following from say scratch buffer to set a prefix key

(global-set-key (kbd "C-c m") 'catmacs-keymap)

And it works fine. I can invoke functions via C-c m prefix (e.g.: C-c m f for setting frequency)

I have a couple of questions.

Q1. Am I doing the prefix command part correctly when catmacs-mode is invoked, and
Q2. What is best practice for getting end user to issue their version of the  global-set-key invocation I show above and where
      Is the best place to document it. For completeness I could place a comment in the code for users who look at the
     Lisp, but for other end users, do I just say execute this or ???


Thanks / Frank

pymaximus commented:
> On 8 Aug 2017, at 2:04 pm, Frank Singleton <b17flyboy@gmail.com> wrote:
> 
> Hi,
> 
> Following up on the keymap and prefix idea (which is rather new to me)..
> 
>>> 	• C-c m is reserved for end-user customisation and should not be used. Instead, choose an allowed alternative or (better) provide a prefix map, and tell the user how to bind it to their prefix key of choice.
> 
> I defined a prefix map as follows ..
> 

Or is this approach preferred ??

(defvar catmacs-prefix "C-c m”) <=====

(define-minor-mode catmacs-mode
  "CAT client for Yaesu FT991A Transceiver"
  :lighter " catmacs"
  :keymap (let ((catmacs-keymap (make-sparse-keymap)))

            ;; keys
            (define-key catmacs-keymap (kbd (concat catmacs-prefix "f")) 'catmacs-fa-set)
            (define-key catmacs-keymap (kbd (concat catmacs-prefix "m")) 'catmacs-md-set)
            (define-key catmacs-keymap (kbd (concat catmacs-prefix "q")) 'catmacs-qr-set)

Guidance on this appreciated :-)


> (define-minor-mode catmacs-mode
>  "CAT client for Yaesu FT991A Transceiver"
>  :lighter " catmacs"
>  :keymap (let ((catmacs-keymap (make-sparse-keymap)))
>            ;; prefix command
>            (define-prefix-command 'catmacs-keymap). <======***
>            ;; keys
>            (define-key catmacs-keymap (kbd "f") 'catmacs-fa-set)
>            (define-key catmacs-keymap (kbd "m") 'catmacs-md-set)
>              <snip>
> 
> 
> I can invoke the following from say scratch buffer to set a prefix key
> 
> (global-set-key (kbd "C-c m") 'catmacs-keymap)
> 
> And it works fine. I can invoke functions via C-c m prefix (e.g.: C-c m f for setting frequency)
> 
> I have a couple of questions.
> 
> Q1. Am I doing the prefix command part correctly when catmacs-mode is invoked, and
> Q2. What is best practice for getting end user to issue their version of the  global-set-key invocation I show above and where
>      Is the best place to document it. For completeness I could place a comment in the code for users who look at the
>     Lisp, but for other end users, do I just say execute this or ???
> 
> 
> Thanks / Frank

purcell commented:
The former approach would be preferable (ie. with `define-prefix-command`).

pymaximus commented:
OK, just committed that change to catmacs repo.

> On 26 Aug 2017, at 11:36 am, Steve Purcell <notifications@github.com> wrote:
> 
> The former approach would be preferable (ie. with define-prefix-command).
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/melpa/melpa/pull/4912#issuecomment-325070716>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AA-yZx7rbMvbEHuJEOrLPu20qpBOJVK-ks5sb3aUgaJpZM4OsR_T>.
>

purcell commented:
Great, merged now. Welcome to MELPA!

pymaximus commented:
> On 27 Aug 2017, at 4:05 pm, Steve Purcell <notifications@github.com> wrote:
> 
> Great, merged now. Welcome to MELPA!
> 
> 
Awesome, thanks !!
